trait system refactor initiative update (#1120)
rust-lang/blog.rust-lang.org: posts/inside-rust/2023-07-11-trait-system-refactor-initiative.md

---
layout: post
title: "Rustc Trait System Refactor Initiative Update"
author: lcnr
team: The Rustc Trait System Refactor Initiative <https://github.com/rust-lang/trait-system-refactor-initiative/>
---

As announced in the [Types Team announcement post](https://blog.rust-lang.org/2023/01/20/types-announcement.html) at the start of this year, the Types Team has started to reimplement the trait solver of rustc. This refactor is similar to [chalk](https://github.com/rust-lang/chalk/), but directly integrated into the existing codebase using the experience gathered over the last few years. Unlike chalk, the new trait solver has the sole goal of replacing the exisiting implementation. We are separately working on formalizing the type system in [a-mir-formality](https://github.com/rust-lang/a-mir-formality).

> By reimplementing the trait solver of rustc we are able to fix many subtle bugs and inefficiencies of the existing implementation. This should result in faster compilation speed and bugs. The new trait solver implementation should also unblock many future changes, most notably around implied bounds and coinduction. This will allow us to remove many of the [current restrictions on GATs](https://github.com/rust-lang/rust/issues/91693) and to fix many long-standing unsound issues, for example [#25860](https://github.com/rust-lang/rust/issues/25860). Some unsound issues will already be fixed at the point of stabilization while others will require additional work afterwards. The new solver will also enable or greatly simplify other, still experimental type system extensions, e.g. [generic const expressions](https://github.com/rust-lang/rust/issues/76560) and [non-lifetime binders](https://github.com/rust-lang/rust/issues/108185).

Reviewer: > This should result in faster compilation speed and bugs.

so you get faster bugs?!

Reviewer 2: I assume this is supposed to say "and less bugs" :rofl:

It has now been half a year since that announcement which matches the first step of [our roadmap][roadmap].

## The status quo

The new trait solver implementation can be tested on nightly by using the rustc flag `-Ztrait-solver=next`. To only use the new implementation for coherence, `-Ztrait-solver=next-coherence` can be used. See the current implementation progress of the new trait solver in [its tracking issue](https://github.com/rust-lang/rust/issues/107374).

We are now at a point where the we exclusively rely on the new implementation when the feature flag is enabled. This is a major step as we've previously still relied on the old solver for ["deep normalization"](https://github.com/rust-lang/rust/pull/113086) and ["selection"](https://github.com/rust-lang/rust/pull/112869). When using the new trait solver many crates and most of our existing regression tests successfully compile.

> While there is a significant tail of less common bugs, we currently have two main failure causes. The new solver has a slightly different approach for `impl Trait`, the implementation of which is still broken for nested RPIT, e.g. `fn foo() -> impl Trait<Assoc = impl Sized>`. Implementing this new approach helped us to discover issues of the existing implementation which allows us to refine its design before the stabilization of TAIT. The inference of implicit unsize coercions, e.g. converting `Box<String>` to `Box<dyn Display>`, relies on implementation details of the existing trait solver. We've recently started to emulating the existing behavior here and should fix most of the remaining breakage from that over the next few weeks.

Reviewer 3: 1. While an example is given for what the `RPIT` acronym means, the acronym itself (return position impl trait, I presume) is not actually defined and only used once. Perhaps the acronym is unnecessary?
2. Same comment for `TAIT`, except in this case I do not know what this acronym stands for, so I don't have a concrete suggestion to fix the text.

```suggestion
While there is a significant tail of less common bugs, we currently have two main failure causes. The new solver has a slightly different approach for `impl Trait`, the implementation of which is still broken for instances of nested return position impl trait, e.g. `fn foo() -> impl Trait<Assoc = impl Sized>`. Implementing this new approach helped us to discover issues of the existing implementation which allows us to refine its design before the stabilization of TAIT (???). The inference of implicit unsize coercions, e.g. converting `Box<String>` to `Box<dyn Display>`, relies on implementation details of the existing trait solver. We've recently started to emulating the existing behavior here and should fix most of the remaining breakage from that over the next few weeks.
```

## Going forward

After fixing the currently open issues, we intend to move parts of rustc to the new trait solver implementation in steps, starting by using it in coherence. We expect to move coherence to the new implementation at the end of this year. Moving the type checking of functions to the new trait solver implementation will be a lot more challenging than coherence. This will be the last place where we will use the old implementation. We expect to change the default there in 2024, potentially relying on the new edition to help with migration.

> A major challenge will be "incompleteness". The trait system sometimes unnecessarily guides type inference. Consider the following example:

Reviewer 2: ```suggestion
A major challenge will be "incompleteness" (note the usage of "incompleteness" here is not the colloquial definition but instead a term of art with a precise technical meaning). The trait system sometimes unnecessarily guides type inference. Consider the following example:
```

Reviewer 3: This leaves me wondering what the technical meaning of "incompleteness" is. Is there a simplified summary that could be included for those of us who are curious?

```suggestion
A major challenge will be "incompleteness". In this context, incompleteness is a technical term which means <short explanation>. The trait system sometimes unnecessarily guides type inference. Consider the following example:
```

Author: updated, does this explanation look useful?

Reviewer 3: Yes, thank you!
